auth0/src/main/java/com/auth0/android/authentication/storage/CryptoUtil.java

```diff
@@ -230,31 +329,79 @@ private void deleteKeys() {
             storage.store(KEY_ALIAS, encodedEncryptedAESText);
             return aes;
         } catch (NoSuchAlgorithmException e) {
+            /*
+             * This exceptions are safe to be ignored:
+             *
+             * - NoSuchAlgorithmException:
+             *      Thrown if the Algorithm implementation is not available. AES was introduced in API 1
+             *
+             * However if any of this exceptions happens to be thrown (OEMs often change their Android distribution source code),
+             * all the checks performed in this class wouldn't matter and the device would not be compatible at all with it.
+             *
+             * Read more in https://developer.android.com/reference/javax/crypto/KeyGenerator
+             */
             Log.e(TAG, "Error while creating the AES key.", e);
-            throw new KeyException("Error while creating the AES key.", e);
+            throw new IncompatibleDeviceException(e);
         }
     }
 
-    //Only used to decrypt final DATA
-    public byte[] decrypt(byte[] encryptedInput) throws CryptoException {
+
+    /**
+     * Encrypts the given input bytes using a symmetric key (AES).
+     * The AES key is stored protected by an asymmetric key pair (RSA).
+     *
+     * @param encryptedInput the input bytes to decrypt. There's no limit in size.
+     * @return the decrypted output bytes
+     * @throws CryptoException             if the RSA Key pair was deemed invalid and got deleted. Operation can be retried.
+     * @throws IncompatibleDeviceException in the event the device can't understand the cryptographic settings required
+     */
+    public byte[] decrypt(byte[] encryptedInput) throws CryptoException, IncompatibleDeviceException {
         try {
             SecretKey key = new SecretKeySpec(getAESKey(), ALGORITHM_AES);
             Cipher cipher = Cipher.getInstance(AES_TRANSFORMATION);
             String encodedIV = storage.retrieveString(KEY_IV_ALIAS);
             if (TextUtils.isEmpty(encodedIV)) {
-                throw new InvalidAlgorithmParameterException("The AES Key exists but an IV was never stored. Try to encrypt something first.");
+                //AES key was JUST generated. If anything existed before, should be encrypted again first.
+                throw new CryptoException("The encryption keys changed recently. You need to re-encrypt something first.", null);
```

Comment:
This is a special scenario where the keys got regenerated but nothing was saved after that.. since we're encrypting with AES in a symmetric manner, the previous stored initialization vector (iv) is no longer related to the now regenerated AES keys. Given that, anything that was already stored cannot be decrypted and should be ignored/removed. The exception message instructs the dev to call "save" credentials first before trying to call "get" again.